lib/chat_models/chat_anthropic.ex

```diff
@@ -0,0 +1,560 @@
+                  # list of lists of deltas.
+                  result = [result]
+                  {processed_data ++ [result], incomplete}
+              end
+            end)
+          end
+        )
+    )
+    |> case do
+      {:ok, %Req.Response{body: data}} ->
+        data
+
+      {:error, %LangChainError{message: reason}} ->
+        {:error, reason}
+
+      {:error, %Mint.TransportError{reason: :timeout}} ->
+        {:error, "Request timed out"}
+
+      {:error, %Mint.TransportError{reason: :closed}} ->
+        # Force a retry by making a recursive call decrementing the counter
+        Logger.debug(fn -> "Mint connection closed: retry count = #{inspect(retry_count)}" end)
+        do_api_request(anthropic, messages, functions, callback_fn, retry_count - 1)
+
+      other ->
+        Logger.error(
+          "Unhandled and unexpected response from streamed post call. #{inspect(other)}"
+        )
+
+        {:error, "Unexpected response"}
+    end
+  end
+
+  defp headers(api_key, api_version) do
+    %{
+      "x-api-key" => api_key,
+      "content-type" => "application/json",
+      "anthropic-version" => api_version
+    }
+  end
+
+  # Parse a new message response
+  @doc false
+  @spec do_process_response(data :: %{String.t() => any()} | {:error, any()}) ::
+          Message.t()
+          | [Message.t()]
+          | MessageDelta.t()
+          | [MessageDelta.t()]
+          | {:error, String.t()}
+  def do_process_response(%{
+        "type" => "message",
+        "content" => [%{"type" => "text", "text" => content}],
+        "stop_reason" => stop_reason
+      }) do
+    %{
+      role: :assistant,
+      content: content,
+      status: stop_reason_to_status(stop_reason)
+    }
+    |> Message.new()
+    |> to_response()
+  end
+
+  def do_process_response(%{
+        "type" => "content_block_start",
+        "content_block" => %{"type" => "text", "text" => content}
+      }) do
+    %{
+      role: :assistant,
+      content: content,
+      status: :incomplete
+    }
+    |> MessageDelta.new()
+    |> to_response()
+  end
+
+  def do_process_response(%{
+        "type" => "content_block_delta",
+        "delta" => %{"type" => "text_delta", "text" => content}
+      }) do
+    %{
+      role: :assistant,
+      content: content,
+      status: :incomplete
+    }
+    |> MessageDelta.new()
+    |> to_response()
+  end
+
+  def do_process_response(%{
+        "type" => "message_delta",
+        "delta" => %{"stop_reason" => stop_reason}
+      }) do
+    %{
+      role: :assistant,
+      content: "",
+      status: stop_reason_to_status(stop_reason)
+    }
+    |> MessageDelta.new()
+    |> to_response()
+  end
+
+  # These streaming events don't contain any useful data to return to callers. Ignore.
+  def do_process_response(%{"type" => type})
+      when type in ["message_start", "ping", "content_block_stop", "message_stop"] do
+    :ignore
```

Comment:
Is this being used? It's not part of the spec and doesn't really fit in with the other return types. Wondering if it's actively being used here or if there's another way to convey this information.

Reply:
With the way the code is currently organized, yes this clause would be needed when streaming responses.

Here's a curl request with its response you can use to see these values:

```shell
curl https://api.anthropic.com/v1/messages \
--header "x-api-key:  $ANTHROPIC_API_KEY" \
--header "content-type: application/json" \
--header "anthropic-version: 2023-06-01" \
--data \
'{
    "model": "claude-3-opus-20240229",
    "max_tokens": 1024,
    "messages": [
        {"role": "user", "content": "Hello, world"}
    ],
    "stream": true
}'
```

```
event: message_start
data: {"type":"message_start","message":{"id":"msg_01F8kFTQzuVBK3V53N36vZGN","type":"message","role":"assistant","content":[],"model":"claude-3-opus-20240229","stop_reason":null,"stop_sequence":null,"usage":{"input_tokens":10,"output_tokens":1}}}

event: content_block_start
data: {"type":"content_block_start","index":0,"content_block":{"type":"text","text":""}               }

event: ping
data: {"type": "ping"}

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":"Hello"} }

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":"!"} }

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":" How"}              }

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":" can"}      }

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":" I"}}

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":" assist"}         }

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":" you"}           }

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":" today"}  }

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"text_delta","text":"?"}    }

event: content_block_stop
data: {"type":"content_block_stop","index":0         }

event: message_delta
data: {"type":"message_delta","delta":{"stop_reason":"end_turn","stop_sequence":null},"usage":{"output_tokens":12}     }

event: message_stop
data: {"type":"message_stop"    }

```

Now, we could change the parser logic below to try and ignore these events rather than allowing them to bubble up to this function, but to me that felt like an awkward abstraction because it would mean that the code responsible for string parsing is also aware of what events have significance to the caller. The original intent was to have the string parsing be its own concern and leave any high-level filtering logic up to callers.

That said, this `:ignore` pattern here is less than ideal so I don't feel strongly. Happy to rework a bit

Reply:
Thanks for putting that all into context here. With the event information always included in the data as "type", is there any reason to pay attention to the event on the outside? 

Wondering if we always grab the data and parse it to JSON, and ingore the "event" part of the message. Wouldn't that simplify the parsing? Is there another reason to parse the the outer event?

Reply:
At the moment, no reason to keep the event that I can think of. My line of thinking was to keep the parser pure in the sense that it just returns objects representing the raw stream, but that's just a preference. Since there's no real need, I removed those.